pkg/reconciler/managed/metrics.go

```diff
@@ -0,0 +1,204 @@
+		Name:      "resource_drift_seconds",
+		Help:      "ALPHA: How long since the previous successful reconcile when a resource was found to be out of sync; excludes restart of the provider",
+		Buckets:   kmetrics.ExponentialBuckets(10e-9, 10, 10),
+	}, []string{"group", "kind"})
+
+	mr = prometheus.NewGaugeVec(prometheus.GaugeOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "managed_resource_created",
+		Help:      "The number of managed resources created",
+	}, []string{"gvk", "name", "claim", "composite"})
+
+	mrReady = prometheus.NewGaugeVec(prometheus.GaugeOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "managed_resource_ready",
+		Help:      "The number of managed resources in Ready=True state",
+	}, []string{"gvk", "name", "claim", "composite"})
+
+	mrReadyDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "managed_resource_ready_duration_seconds",
+		Help:      "The time it took for a managed resource to become ready first time after creation",
+		Buckets:   []float64{1, 5, 10, 15, 30, 60, 120, 300, 600, 1800, 3600},
+	}, []string{"gvk", "name", "claim", "composite"})
+
+	mrDetected = prometheus.NewHistogramVec(prometheus.HistogramOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "managed_resource_detected_time_seconds",
+		Help:      "The time it took for a managed resource to be detected by the controller",
+		Buckets:   kmetrics.ExponentialBuckets(10e-9, 10, 10),
+	}, []string{"gvk", "name", "claim", "composite"})
+
+	mrSynced = prometheus.NewGaugeVec(prometheus.GaugeOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "managed_resource_synced",
+		Help:      "The number of managed resources in Synced=True state",
+	}, []string{"gvk", "name", "claim", "composite"})
+
+	mrDeletionDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "managed_resource_deletion_seconds",
+		Help:      "The time it took for a managed resource to be deleted",
+		Buckets:   []float64{1, 5, 10, 15, 30, 60, 120, 300, 600, 1800, 3600},
+	}, []string{"gvk", "name", "claim", "composite"})
+)
+
+type metricRecorder struct {
+	firstObservation sync.Map
+	lastObservation  sync.Map
+
+	cluster cluster.Cluster
+	gvk     schema.GroupVersionKind
+}
+
+func (r *metricRecorder) Start(ctx context.Context) error {
+	inf, err := r.cluster.GetCache().GetInformerForKind(ctx, r.gvk)
+	if err != nil {
+		return errors.Wrapf(err, "cannot get informer for metric recorder for resource %s", r.gvk)
+	}
+
+	registered, err := inf.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
Will this definitely notice every deleted MR?

Is there a reason to do this cleanup in an edge-driven informer function, rather than doing it immediately after calling `r.metricRecorder.recordDeleted`?